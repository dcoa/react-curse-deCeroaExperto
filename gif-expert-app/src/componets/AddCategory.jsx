import React, { useState } from 'react';
import PropTypes from 'prop-types';


export const AddCategory = ({ setCategory }) => {
    const [inputValue, setinputValue] = useState('');

    const handleSubmit = (e) => {
        // eslint-disable-next-line no-unused-expressions
        e.preventDefault();
        if( inputValue.trim().length > 2){
            setCategory(categ => [inputValue, ...categ])
            setinputValue('')
        }
    }
    return (
        <>
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    value={inputValue}
                    placeholder='Escribe lo que quieres buscar y presiona "Enter"'
                    onChange={e => setinputValue(e.target.value)}
                />
            </form>
        </>
    )
}

AddCategory.propTypes = {
    setCategory: PropTypes.func.isRequired
}
