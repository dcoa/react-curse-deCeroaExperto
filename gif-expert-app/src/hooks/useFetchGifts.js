import { useState, useEffect } from "react"
import { getGifs } from '../helpers/getGifts'
export const useFetchGifts = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
        getGifs(category)
            .then(img => setState({
                data: img,
                loading: false
            }))
        //El arreglo de dependencias esta vacio solo se ejecuta 1 unica vez
        //si cambia el valor de category vuelve y hace la petici√≥n (en este caso el valor no cambia)
    }, [category])

    return state;
}