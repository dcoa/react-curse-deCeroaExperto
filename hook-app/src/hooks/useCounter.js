import { useState } from "react"

export const useCounter = (initialState = 10) => {
    const [counter, setCounter] = useState(initialState);

    // el usuario ingresa el valor en el cual se incrementa el estado
    const increment = (factor) => {
        setCounter(counter + factor)
    };
    // el usuario ingresa el valor en el cual se decrementa el estado
    const decrement = (factor) => {
        setCounter(counter - factor)
    };
    // el usuario reestablece el contador
    const reset = () => {
        setCounter(initialState)
    };
    return {
        counter,
        reset,
        increment,
        decrement
    }
}
