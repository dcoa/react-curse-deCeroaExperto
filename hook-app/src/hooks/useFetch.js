import React, { useEffect, useState, useRef } from 'react'

export const useFetch = ( url ) => {
    const isMounted = useRef(true);
    const [state, setstate] = useState({data: null, loading: true, error: null})

         
    useEffect(() => {
        
        return () => {
            isMounted.current = false
        }
    }, [])

    useEffect(() => {
        // Realiza el fetch del componente que llama el use fetch solo si esta montado
        if( isMounted.current ) {
        // limpiar el estado anterior antes de hacer el nuevo llamado al api
        setstate({data: null, loading: true, error: null})
        fetch( url )
        .then( resp => resp.json())
        .then(resp => setstate({error: null, loading: false, data: resp}))
        .catch()
        }

    }, [url])

    return state;
}
