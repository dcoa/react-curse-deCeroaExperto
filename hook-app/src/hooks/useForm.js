import { useState } from 'react'

//sirve para manejar los campos del formulario y leerlos
export const useForm = (initialState = {}) => {
   const [formValues, setformValues] = useState(initialState);
   const reset = () => setformValues(initialState)
   const handleInputChange = ({target}) => {
    setformValues({
        ...formValues,
        [target.name]: target.value
    })
}
return{
    formValues,
    handleInputChange,
    reset
}
}
