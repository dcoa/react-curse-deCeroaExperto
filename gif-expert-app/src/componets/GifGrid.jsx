import React from 'react';
import PropTypes from 'prop-types';
import { useFetchGifts } from '../hooks/useFetchGifts';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ Category }) => {
    // const [images, setImages] = useState([]);
    const { data, loading } = useFetchGifts(Category);
    //transformaci√≥n para crear una galeria de imagenes
    let columns = [];

    for (let i = 0; i < data.length; i += 3) {
        columns.push(data.slice(i, i + 3))
    }

    return (
        <>
            <h3>{Category}</h3>
            <div className='card-grid'>
                {loading && 'Cargandoo...'}
                {
                    columns.map((images, i) => <div key={'columns' + i} className='columns'>
                        {images.map(img => <GifGridItem key={img.id} {...img} />)}
                    </div>)
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    Category: PropTypes.string.isRequired
}
