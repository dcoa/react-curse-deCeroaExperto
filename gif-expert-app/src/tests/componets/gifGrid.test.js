import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { GifGrid } from '../../componets/GifGrid'
import { useFetchGifts } from '../../hooks/useFetchGifts';

jest.mock('../../hooks/useFetchGifts');

describe('Testing in gifGrid.js', () => {
    const cat = 'one piece'

    test('should render correct', () => {
        //mock de el hook inicial
        useFetchGifts.mockReturnValue({
            data: [],
            loading: true
        });
        // no lo puedo colocar por fuera porque no toma el mock
        const wrapper = shallow(<GifGrid Category={cat} />)
        expect(wrapper).toMatchSnapshot();
    })

    test('should show items when images are loaded', () => {
        const img = [{
            id: 'ABC',
            url: 'http://localhost(cualquiercosa/cosa.jpg',
            title: 'Prueba'
        },
        {
            id: 'CDE',
            url: 'http://localhost(cualquiercosa/cosa.jpg',
            title: 'Prueba 2'
        }]
        // Se hace un mock del constum hook useFetchGifs
        useFetchGifts.mockReturnValue({
            data: img,
            loading: false
        });
        const wrapper = shallow(<GifGrid Category={cat} />)
        //Forma mas facil de analizar el renderizado expect(wrapper).toMatchSnapshot()
        //Encnenutra el elemento GifGridItem y cuantos hay en este caso solo debe haber 1 
        expect(wrapper.find('GifGridItem').length).toBe(img.length)
    })

})
