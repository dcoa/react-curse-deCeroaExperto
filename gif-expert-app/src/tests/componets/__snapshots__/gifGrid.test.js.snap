// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in gifGrid.js should render correct 1`] = `
<Fragment>
  <h3>
    one piece
  </h3>
  <div
    className="card-grid"
  >
    Cargandoo...
  </div>
</Fragment>
`;
